app/services/search_service.rb

```diff
@@ -0,0 +1,55 @@
+class SearchService
+
+  def self.sync_search_vector!(discussion_id)
+    SearchVector.where(discussion_id: discussion_id).delete_all
+
+    SearchVector.execute_search_query(
+      SearchAlgorithms.sync_search_vector,
+      id: discussion_id
+    )
+  end
+
+  def self.search_for(query, user: nil, limit: 10)
+    return [] if (visible_ids = visible_results_for(user)).empty?
+
+    SearchVector.execute_search_query(
+      SearchAlgorithms.search(visible_ids),
+      query: query,
+      limit: limit
+    ).map { |result| build_search_result result }
+  end
+
+  private
+
+  def self.visible_results_for(user)
+    Queries::VisibleDiscussions.new(user: user).pluck(:id)
+  end
+
+  def self.build_search_result(result)
```

Comment:
So, now, when we get a result back, we know how relevant it is, but we don't know *which* proposals or comments made it more relevant. So we have to do a separate query to get these things.

This is the slowest part of the process, but it's still faster than the existing solution (I benchmarked it at ~1.1 seconds for a typical search on my prod database)

I'd also imagine there's ways to speed these relevant_motion/comment queries up a bit.